src/net.cpp

```diff
@@ -699,35 +724,41 @@ const uint256& V1TransportDeserializer::GetMessageHash() const
     return data_hash;
 }
 
-CNetMessage V1TransportDeserializer::GetMessage(const CMessageHeader::MessageStartChars& message_start, int64_t time) {
+Optional<CNetMessage> V1TransportDeserializer::GetMessage(int64_t time)
+{
     // decompose a single CNetMessage from the TransportDeserializer
     CNetMessage msg(std::move(vRecv));
 
-    // store state about valid header, netmagic and checksum
-    msg.m_valid_header = hdr.IsValid(message_start);
-    msg.m_valid_netmagic = (memcmp(hdr.pchMessageStart, message_start, CMessageHeader::MESSAGE_START_SIZE) == 0);
-    uint256 hash = GetMessageHash();
-
-    // store command string, payload size
+    // store command string, time, and sizes
     msg.m_command = hdr.GetCommand();
+    msg.m_time = time;
     msg.m_message_size = hdr.nMessageSize;
     msg.m_raw_message_size = hdr.nMessageSize + CMessageHeader::HEADER_SIZE;
 
+    uint256 hash = GetMessageHash();
+
     // We just received a message off the wire, harvest entropy from the time (and the message checksum)
     RandAddEvent(ReadLE32(hash.begin()));
 
-    msg.m_valid_checksum = (memcmp(hash.begin(), hdr.pchChecksum, CMessageHeader::CHECKSUM_SIZE) == 0);
-    if (!msg.m_valid_checksum) {
-        LogPrint(BCLog::NET, "CHECKSUM ERROR (%s, %u bytes), expected %s was %s\n",
-                 SanitizeString(msg.m_command), msg.m_message_size,
-                 HexStr(hash.begin(), hash.begin()+CMessageHeader::CHECKSUM_SIZE),
-                 HexStr(hdr.pchChecksum, hdr.pchChecksum+CMessageHeader::CHECKSUM_SIZE));
+    // Check checksum
+    // Always reset the network deserializer (prepare for the next message)
+    if (memcmp(hash.begin(), hdr.pchChecksum, CMessageHeader::CHECKSUM_SIZE) != 0) {
+        LogPrint(BCLog::NET, "CHECKSUM ERROR (%s, %u bytes), expected %s was %s, peer=%d\n",
+            SanitizeString(msg.m_command), msg.m_message_size,
+            HexStr(hash.begin(), hash.begin() + CMessageHeader::CHECKSUM_SIZE),
+            HexStr(hdr.pchChecksum, hdr.pchChecksum + CMessageHeader::CHECKSUM_SIZE),
+            node_id);
+        Reset();
```

Comment:
5ad7ce1 and 3a9d277 I agree with @jnewbery here about the multiple Resets. Maybe something like this would be cleaner (tested example):
```diff
-    // Check checksum
-    // Always reset the network deserializer (prepare for the next message)
+    Optional<CNetMessage> result{};
     if (memcmp(hash.begin(), hdr.pchChecksum, CMessageHeader::CHECKSUM_SIZE) != 0) {
         LogPrint(BCLog::NET, "CHECKSUM ERROR ... ;
-        Reset();
-        return nullopt;
-    }
-    // Check header command string
-    if (!hdr.IsCommandValid()) {
+    } else if (!hdr.IsCommandValid()) {
         LogPrint(BCLog::NET, "HEADER ERROR - COMMAND ... ;
-        Reset();
-        return nullopt;
+    } else {
+        result = msg;
     }
-
+    // Reset the network deserializer (prepare for the next message).
     Reset();
-    return msg;
+    return result;
```

Reply:
Thanks for the tested example!  I'm looking deeper into how Optional is implemented.  It looks like we don't have move semantics until before Boost 1.56.0.  I'm worried about the performance hit if we're copying a CNetMessage every time.

If I don't find anything important there, then I'll use this suggestion for sure.

Reply:
IIRC `return result;` should always move if it is possible. All recent OS as well as the binaries on the website come with a version of boost greater than 1.56. I think we don't care about reduced performance when someone compiles on a historic OS.

Reply:
See also commit ff9c671b11d40e5d0623eff3dd12e48cbaafb34e

Reply:
Sounds good to me!